### ISSUE TITLE
[2103.00020] Learning Transferable Visual Models From Natural Language Supervision

### ISSUE THREAD

**Issue by riron1206** [riron1206/read_paper#82]
## 概要
https://arxiv.org/abs/2103.00020
OpenAIのCLIP
画像と画像を説明するテキストのペア４億組を学習させた、汎用画像分類モデルCLIP
画像とテキストの対応関係を学習するモデル
“zero-shot”（学習していないクラス）の画像分類ができる
image to textはできない。あくまで分類のみ

## 新規性・差分
**CLIPのすごいところは「画像のベクトル」と「テキストのベクトル」を同じ潜在空間上で扱えるところ**
例えば

- ある画像のベクトルに近いテキストはどれか？
→要は画像分類。与えられた1文に含まれる単語の中で、与えられた1枚の画像に最適な単語（ラベル）を予測する

- あるテキストのベクトルに近い画像のベクトルはどれか？
→要は画像検索。与えられた1文に最適な画像を予測する
<img src=https://user-images.githubusercontent.com/22429441/113414529-2b25c480-93f8-11eb-97c1-330eb5379d6a.png width=500>

といったことができる

## 手法
**1. 学習 (Contrasive pre-training)**
「画像」と「その内容を説明するテキスト」のペアがモデルのinput
↓
画像とテキストをそれぞれのEncoderに通す
↓
画像とテキスト2つのベクトルの内積（COS類似度）計算。大きいほど類似度が高い
↓
内積の行列ができる
↓
対角成分が画像とテキストのペアの成分なので、対角成分を最大化、それ以外の成分を最小化するように学習
↓
画像からもテキストからも、その特徴を適切に表すベクトルを取得できるようになる

![img](http://cedro3.com/wp-content/uploads/2021/03/%E3%83%95%E3%83%AD%E3%83%BC%EF%BC%91-1.png)

**2. 予測**

**2-1. textから分類クラスのベクトル作成（Create dataset classifier from label text）**
画像分類したいクラス名を含むテキスト用意（例えばクラスがdogであれば「A photo of a dog」というような文）。このテキストは学習データに近い画像のキャプションのような構文でないとうまくいかない
↓
テキストをText Encoderへ入力して特徴ベクトル T_1〜T_N を取り出す

**2-2. 画像分類（Use for sero-shot prediction）**
分類したい画像をImage Encoderへ入力して画像の特徴ベクトル I_1 を取り出す
↓
画像の特徴ベクトルとテキストの特徴ベクトルの掛け算を行う
↓
一番大きい値となったクラスをラベルとする
↓
膨大なデータで画像とテキストのペアの特徴を学んでるため、画像のオブジェクトに対応した単語の分類が精度よくできるらしい。**従来のsoftmaxを使った方法のような「dogクラスはこんな画像」と明示的な学習なしで特定クラス分類できるからzero shotと言ってるみたい（metric learningに近い。COS類似度で対応する対角成分のベクトルを最大化するように学習してるし）**

![img](http://cedro3.com/wp-content/uploads/2021/03/%E3%83%95%E3%83%AD%E3%83%BC%EF%BC%92-1.png)

## コメント
github
- https://github.com/OpenAI/CLIP  # 公式
- https://sachinruk.github.io/blog/pytorch/pytorch%20lightning/loss%20function/gpu/2021/03/07/CLIP.html  # pytorch-lightning

参考サイト
- https://openai.com/blog/clip/  # openAIのblog
- https://ganpaint.io/miniclip/  # demoサイト。テキストに対するgrad-camもやってる
- http://cedro3.com/ai/clip-search/  # めっちゃわかりやすい
- https://github.com/yoheikikuta/paper-reading/issues/57
- /content/drive/MyDrive/Work/CLIP_search  # **画像検索colabで試してみた。テキストに近い画像出せるのすごい**
